networks/2_module_validation.py

```diff
@@ -0,0 +1,300 @@
+pao1_regulon["Length_processed"] = pao1_regulon["Genes_processed"].str.len()
+
+# For each regulon/operon, select a random set of genes that are the same size at the regulon/operon
+pao1_operon["Random_Genes"] = pao1_operon["Length_processed"].apply(
+    lambda num_genes: pao1_membership.sample(num_genes).index.values
+)
+pao1_regulon["Random_Genes"] = pao1_regulon["Length_processed"].apply(
+    lambda num_genes: pao1_membership.sample(num_genes).index.values
+)
+
+pao1_operon.head()
+
+pao1_regulon.head()
+
+# ### Calculate the distribution
+
+# For each regulon/operon get the number of modules that regulon/operon genes are found in, number of modules
+# that random genes are found in
+pao1_operon["Num_operon_modules"] = pao1_operon["Genes_processed"].apply(
+    lambda list_genes: pao1_membership.loc[list_genes]["module id"].nunique()
+)
+pao1_operon["Num_random_modules"] = pao1_operon["Random_Genes"].apply(
+    lambda list_genes: pao1_membership.loc[list_genes]["module id"].nunique()
+)
+
+pao1_regulon["Num_regulon_modules"] = pao1_regulon["Genes_processed"].apply(
+    lambda list_genes: pao1_membership.loc[list_genes]["module id"].nunique()
+)
+pao1_regulon["Num_random_modules"] = pao1_regulon["Random_Genes"].apply(
+    lambda list_genes: pao1_membership.loc[list_genes]["module id"].nunique()
+)
+
+pao1_operon.head()
+
+pao1_regulon.head()
+
+# +
+# Format df for plotting using displot
+pao1_operon_toplot = pd.melt(
+    pao1_operon, value_vars=["Num_operon_modules", "Num_random_modules"]
+)
+pao1_regulon_toplot = pd.melt(
+    pao1_regulon, value_vars=["Num_regulon_modules", "Num_random_modules"]
+)
+
+pao1_operon_toplot.tail()
+
+
+# -
+
+def cumulative_distribution(
+    data,
+    scaled=False,
+    survival=False,
+    label="Cumulative",
+    fill=False,
+    flip=False,
+    preserve_ends=0,
+    **kwargs,
+):
+    """
+    plots cumulative (or survival) step distribution
+    adapted from https://github.com/MarvinT/morphs/blob/master/morphs/plot/utils.py
+    """
+    data = np.sort(data)
+    if survival:
+        data = data[::-1]
+    y = np.arange(data.size + 1, dtype=float)
+    if scaled:
+        y /= y[-1]
+    x = np.concatenate([data, data[[-1]]])
+    plt.step(x, y, label=label, **kwargs)
+    if fill:
+        plt.fill_between(x, y, alpha=0.5, step="pre", **kwargs)
+
+
+# +
+cumulative_distribution(
+    pao1_operon["Num_operon_modules"],
+    label="Number of modules containing operon genes",
+    color="red",
+)
+cumulative_distribution(
+    pao1_operon["Num_random_modules"],
+    label="Number of modules containing random genes",
+    color="blue",
+)
+_ = plt.legend()
+plt.title("Distribution of module counts (operon vs random genes)")
+
+scipy.stats.ks_2samp(
+    pao1_operon["Num_operon_modules"], pao1_operon["Num_random_modules"]
+)
+# -
+
+fig = sns.displot(
+    pao1_operon_toplot,
+    x="value",
+    hue="variable",
+)
+plt.title("Distribution of module counts (operon vs random genes)")
```

Comment:
So is this the same as above, but a PMF (not cumulative) instead of a CDF (cumulative)?

Reply:
Yes these are a couple different visualizations of the same data. I'll update the titles to be more specific